File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t) :: map
  def count(sentence) do
    sentence
    |> sanitize
    |> String.split
    |> Enum.reduce(%{}, fn x, acc -> Map.update(acc, x, 1, &(&1 + 1)) end)
  end

  defp sanitize(sentence) do
    sentence
    |> String.downcase
    |> String.replace("_", " ") # Underscores are spaces.
    |> String.replace(~r/[^\p{L}\d -]/iu, "") # Keep only letters, numbers, spaces, and hyphens.
  end
end

File: ecmascript/word-count/word-count.js
class Words {
  counter(acc, item) {
    if (!acc.hasOwnProperty(item)) { acc[item] = 0; }
    acc[item] += 1;
    return acc;
  };

  count(words) {
    return words
      .trim()
      .toLowerCase()
      .split(/\s+/)
      .reduce(this.counter, {});
  }
}

export default Words;